chainstate/tx-verifier/src/transaction_verifier/signature_destination_getter.rs

```diff
@@ -233,8 +245,44 @@ impl<'a> SignatureDestinationGetter<'a> {
         Self { f }
     }
 
-    pub fn call(&self, input: &TxInput) -> Result<Destination, SignatureDestinationGetterError> {
-        (self.f)(input)
+    pub fn call(
+        &self,
+        input: &TxInput,
+        witness: &InputWitness,
+    ) -> Result<Destination, SignatureDestinationGetterError> {
+        (self.f)(input, witness)
+    }
+}
+
+fn get_destination_from_htlc(
+    spending_time: &BlockTimestamp,
+    input_witness: &InputWitness,
+    htlc: &HashedTimelockContract,
+    outpoint: &UtxoOutPoint,
+) -> Result<Destination, SignatureDestinationGetterError> {
+    match input_witness {
+        InputWitness::NoSignature(_) => todo!(),
+        InputWitness::Standard(sig) => {
+            let destination = match AuthorizedPublicKeyHashSecretSpend::from_data(
+                sig.raw_signature(),
+            ) {
+                Ok(_) => htlc.spend_key,
+                Err(_) => {
+                    match htlc.refund_timelock {
+                        OutputTimeLock::UntilHeight(_)
+                        | OutputTimeLock::ForBlockCount(_)
+                        | OutputTimeLock::ForSeconds(_) => {
+                            return Err(SignatureDestinationGetterError::HashedTimelockContractUnsupportedTimelock);
+                        }
+                        OutputTimeLock::UntilTime(t) => {
+                            ensure!(spending_time <= t, SignatureDestinationGetterError::HashedTimelockContractTimelockViolation(outpoint.clone()));
```

Comment:
This check shouldn't be done when getting the destination. It should be done in the spend logic.